lib/byron/src/Cardano/Wallet/Byron/Network.hs

```diff
@@ -676,75 +698,114 @@ chainSyncWithBlocks
             }
 
-        clientStNext
-            :: ([ByronBlock], Int)
-            -> (NextBlocksResult (Point ByronBlock) ByronBlock -> m ())
-            -> ClientStNext ByronBlock (Tip ByronBlock) m Void
-        clientStNext (blocks, n) respond
-            | n <= 1 = ClientStNext
-                { recvMsgRollBackward = onRollback
-                , recvMsgRollForward = \block tip ->
-                    ChainSyncClient $ do
-                        let cursor' = blockPoint block
-                        let blocks' = reverse (block:blocks)
-                        let tip'    = fromTip getGenesisBlockHash getEpochLength tip
-                        respond (RollForward cursor' tip' blocks')
-                        clientStIdle
-                }
-            | otherwise = ClientStNext
-                { recvMsgRollBackward = onRollback
-                , recvMsgRollForward = \block _ ->
-                    ChainSyncClient $ pure $ SendMsgRequestNext
-                        (clientStNext (block:blocks,n-1) respond)
-                        (pure $ clientStNext (block:blocks,1) respond)
-                }
-          where
-            onRollback point _tip = ChainSyncClient $ do
+        -- Simple strategy that sends a request and wait for the answer.
+        oneByOne
+            :: RequestNextStrategy m 'Z
+        oneByOne respond = P.SendMsgRequestNext
+            (collectResponses respond [] Zero)
+            (pure $ collectResponses respond [] Zero)
+
+        -- We only pipeline requests when it's safe to pipeline them. Safe
+        -- is when all blocks are considered 'stable' and immutable, i.e.
+        -- when we land more than `k` blocks from the tip, so when we start
+        -- more than `k + maxInFlight` blocks from the tip
+        --
+        --     0                       tip - k           tip
+        --     |---------------------------|-------------->
+        --               pipelined            one by one
+        --
+        pipeline
+            :: Nat n
+            -> RequestNextStrategy m n
+        pipeline (Succ n) respond | natToInt (Succ n) == fromIntegral maxInFlight =
+            P.CollectResponse Nothing $ collectResponses respond [] n
+        pipeline n respond =
+            P.SendMsgRequestNextPipelined $ pipeline (Succ n) respond
+
+        collectResponses
+            :: (NextBlocksResult (Point ByronBlock) ByronBlock -> m ())
+            -> [ByronBlock]
+            -> Nat n
+            -> P.ClientStNext n ByronBlock (Tip ByronBlock) m Void
+        collectResponses respond blocks Zero = P.ClientStNext
+            { P.recvMsgRollForward = \block tip -> do
+                let cursor' = blockPoint block
+                let blocks' = reverse (block:blocks)
+                let tip'    = fromTip getGenesisBlockHash getEpochLength tip
+                respond (RollForward cursor' tip' blocks')
+                distance <- tipDistance (blockNo block) <$> getNodeTip
```

Comment:
Can't you use `tip`, instead of `getNodeTip`?

Reply:
Actually yes. I went through several refactors and trials on this one, so this is a leftover :+1:

Reply:
Fixed https://github.com/input-output-hk/cardano-wallet/compare/c1ef56bb5e3b9c924673981ee6d4ebb36248e3d4..0c1a920da3563ca8cf09e37d620b78b653f17555